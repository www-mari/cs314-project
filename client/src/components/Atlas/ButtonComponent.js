import React, { Component } from 'react';
import {Button} from 'reactstrap';
import 'leaflet/dist/leaflet.css';

export default class ButtonComponent extends Component {

    constructor(props) {
        super(props);
        this.state = {
        }
    }

    render() {
        return (
            <Button
                id={this.props.id}
                color={this.props.color}
                onClick={this.props.onClick}
                className={this.props.className}
            >
                {this.props.text}
            </Button>
        );
    }
}
