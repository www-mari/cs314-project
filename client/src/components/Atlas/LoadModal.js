import React, {Component} from 'react';
import {Row, Input, Modal, ModalHeader, ModalBody, ModalFooter} from 'reactstrap';
import 'leaflet/dist/leaflet.css';
import {latitudeLongitudestoLatLngs, schemaChecker} from '../../utils/fileHandler'

export default class LoadModal extends Component {

    constructor(props) {
        super(props);
        this.hideLoadModal = this.props.hideLoadModal.bind(this);
        this.addLocation = this.props.addLocation.bind(this);
        this.loadFile = this.loadFile.bind(this);
        this.onloadEvent = this.onloadEvent.bind(this);
        this.state={
            fileData: "",
            locations: this.props.locations,
            loadLocations: [], 
        }
    }

    render() {
        return (
            <Modal toggle={this.props.hideLoadModal} isOpen={this.props.isOpen} className="">
                <ModalHeader toggle={this.props.hideLoadModal}>Upload File</ModalHeader>
                <ModalBody>
                    {this.renderFileUploadButton()}
                </ModalBody>
                <ModalFooter>
                </ModalFooter>
            </Modal>
        );
    }

    renderFileUploadButton(){
        return(
            <Row> 
                <Row className="ml-2"> 
                    <Input type="file" name="file" accepts='.json, .csv' id="load-file" onChange={this.loadFile}/>
                    Accepts valid .json and .csv filetypes only
                </Row>
            </Row>
        );
    }

    loadFile(event){
        let extension = event.target.files[0].name;
        extension = extension.substring(extension.length -4);
        let reader = new FileReader();
        reader.onload = () => this.onloadEvent(extension, reader);
        reader.readAsText(event.target.files[0]);
    }

    onloadEvent(extension, reader){
        let data = null;
        let isValidSchema = true;
        if(extension === '.csv'){
            data = this.csvToJsonConverter(reader.result);
        }
        else{
            data = JSON.parse(String("" + reader.result));
            isValidSchema = schemaChecker(data);
        }
        if(isValidSchema){
            if(data.places){
                data = data.places;
            }     
            data = latitudeLongitudestoLatLngs(data);
            this.props.clearLocations([]);
            this.setState({locations: data});
            Object.values(this.state.locations).forEach( location =>
            {
                let notes = "";
                let address = "";
                if(location.notes){
                    notes = location.notes;
                }
                if(location.address){
                    address = location.address.replaceAll(";", ",");
                }
                this.addLocation({latlng:{lat:location.lat,lng:location.lng}, notes: notes, address: address});
            })
            this.props.hideLoadModal();
        }
    }

    //method inspired by http://techslides.com/convert-csv-to-json-in-javascript
    csvToJsonConverter(csv){
        let lines=csv.split("\n");
        let result = [];
        let headers=lines[0].split(",");
        for(let i=1;i<lines.length;i++){
            let obj = {};
            let currentLine=lines[i].split(",");
            if(currentLine.length <= 1){
                continue;
            }
            for(var j=0;j<headers.length;j++){
                obj[headers[j]] = currentLine[j];
            }
            result.push(obj);
        }
        return result;
    }
}
