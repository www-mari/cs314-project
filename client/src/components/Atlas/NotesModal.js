import React, {Component} from 'react';
import {Container, InputGroup, Input, InputGroupAddon, Row, Col, Button, Modal, ModalHeader, ModalBody, ModalFooter, ButtonGroup, Dropdown, Table} from 'reactstrap';
import 'leaflet/dist/leaflet.css';
import {downloadCsv, downloadJson, downloadKml, downloadSvg} from "../../utils/fileHandler";
import {getLocation, buildAddress, editAddress, addNotesToLocation} from '../../utils/geocoding';


export default class NotesModal extends Component {

    constructor(props) {
        super(props);
        this.saveNotesToLocation = this.saveNotesToLocation.bind(this);
        this.editInputAddress = this.editInputAddress.bind(this);
        this.setNotes = this.setNotes.bind(this);
        this.saveNewAddress = this.saveNewAddress.bind(this);
        this.setAddress = this.setAddress.bind(this);

        this.state={
            dropDownOpen: false,
            edit: false,
            notes: null,
            address: this.printAddress()
        }
    }

    render() {
        return (
            <Modal toggle={this.props.hideNotesModal} isOpen={this.props.isOpen} className="">
                <ModalHeader toggle={this.props.hideNotesModal}>Notes: </ModalHeader>
                <ModalBody>
                    <Row className="mb-2">
                        <Col>
                            Current Location:{" "}
                            {this.printAddress()}
                        </Col>
                    </Row>
                    <Row className="mb-2">
                        <Col>
                            <Button id='editNotes' color="primary" onClick={(() => this.editInputAddress())}>Edit Existing Address</Button>
                            {this.renderEditAddress()}
                        </Col>
                    </Row>
                    <Row className="mb-2">
                        <Col>
                            {this.renderNotesImput()}
                            {this.renderNotes()}
                        </Col></Row>
                </ModalBody> <ModalFooter> </ModalFooter>
            </Modal>
        );}

    renderEditAddress(){
        if(this.state.edit){
            return(
                <InputGroup className="mt-2">
                    <Input
                        //placeholder="Edit Existing Address"
                        onKeyPress={event => {
                            if (event.key === "Enter") {
                                this.saveNewAddress(this.state.address);
                            }
                        }}
                        value={this.state.address}
                        onChange={this.setAddress}
                        id="notes-input"/>
                    <InputGroupAddon addonType="append">
                        <Button id='submitNotes' color="primary" onClick={(() => this.saveNewAddress(this.state.address))}>Save</Button>
                    </InputGroupAddon> 
                </InputGroup> 
            );
        }
        else{
            this.state.address = this.printAddress();
        }
    }

    renderNotesImput(){
        return(
            <InputGroup className="mb-2">
                <Input
                    onKeyPress={event => {
                        if (event.key === "Enter") {
                            this.saveNotesToLocation(this.state.notes);
                        }
                    }}
                    placeholder="Add Notes to this location"
                    value={this.state.notes}
                    onChange={this.setNotes}
                    id="notes-input"/>
                <InputGroupAddon addonType="append">
                    <Button id='submitNotes' color="primary" onClick={(() => this.saveNotesToLocation(this.state.notes))}>Submit</Button>
                </InputGroupAddon> 
            </InputGroup> 
        );
    }

    renderNotes(){
        if(this.props.location != null){ //eventually turn into this.props.location.notes
            if(this.props.location.notes != null){
                return(
                    <Row>
                        <Col>
                            Notes: {this.props.location.notes}
                        </Col>
                    </Row>
                );
            }
        }
    }

    editInputAddress(){
        this.setState({edit: true});
    }

    saveNotesToLocation(notesString){
        addNotesToLocation(this.props.location, notesString);
        this.setState({notes: null});
    }

    setNotes(onChangeEvent){
        const inputText = onChangeEvent.target.value;
        this.setState({notes: inputText});
    }
    
    saveNewAddress(newAddress){
        editAddress(this.props.location, newAddress);
        this.setState({edit: false});
        this.setState({address: null});
    }

    setAddress(onChangeEvent){
        const inputText = onChangeEvent.target.value;
        this.setState({address: inputText});
    }

    printAddress(){
        if(this.props.location != null){
            return(this.props.location.address)
        }
        else{
            return null;
        }
    }

    
}
