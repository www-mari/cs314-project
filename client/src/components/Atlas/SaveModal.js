import React, {Component} from 'react';
import {Container, Row, Col, Button, Modal, ModalHeader, ModalBody, ModalFooter, ButtonGroup, Dropdown, Table} from 'reactstrap';
import 'leaflet/dist/leaflet.css';
import {downloadCsv, downloadJson, downloadKml, downloadSvg} from "../../utils/fileHandler";

export default class SaveModal extends Component {

    constructor(props) {
        super(props);
        this.downloadFile = this.downloadFile.bind(this);
        this.state={
            dropDownOpen: false
        }
    }

    render() {
        return (
            <Modal toggle={this.props.hideSaveModal} isOpen={this.props.isOpen} className="">
                <ModalHeader toggle={this.props.hideSaveModal}>Save Options</ModalHeader>
                <ModalBody>
                    {this.formatOptions()}
                </ModalBody>
                <ModalFooter>
                </ModalFooter>
            </Modal>
        );
    }

    formatOptions(){
        return(
            <Row> 
                <Row className="ml-1"> 
                    Please choose what part of your trip you want to save and in what corresponding format you want: 
                </Row>
                <ButtonGroup vertical className="ml-1"> 
                    Tour
                    <ButtonGroup>
                        <Button id="DownloadCSV" size="sm" color="info" onClick={() => this.downloadFile(downloadCsv)}>CSV</Button>
                        <Button id="DownloadJSON" size="sm" color="info" onClick={() => this.downloadFile(downloadJson)}>JSON</Button>
                    </ButtonGroup>
                
                    Map
                    <ButtonGroup>
                        <Button id="DownloadSVG" size="sm" color="info" onClick={() => this.downloadFile(downloadSvg)}>SVG</Button>
                        <Button id="DownloadKML" size="sm" color="info" onClick={() => this.downloadFile(downloadKml)}>KML</Button>
                    </ButtonGroup>
                </ButtonGroup>
            </Row>
        );
    }

    downloadFile(downloadFileType){
        downloadFileType(this.props.locations);
        this.props.hideSaveModal();
    }
}
