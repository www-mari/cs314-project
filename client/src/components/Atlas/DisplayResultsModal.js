import React, {Component} from 'react';
import {
    FormGroup,
    Button,
    Modal,
    ModalHeader,
    ModalBody,
    ModalFooter,
    Table,
    Label,
    Input,
    InputGroupAddon,
    InputGroup,
    Row,
    Container
} from 'reactstrap';
import 'leaflet/dist/leaflet.css';
import ButtonComponent from "./ButtonComponent";

export default class DisplayResultsModal extends Component {

    constructor(props) {
        super(props);
        this.setFilterResultsInputString = this.setFilterResultsInputString.bind(this);
        this.filterResults = this.filterResults.bind(this);
        this.createFilterTermsRow = this.createFilterTermsRow.bind(this);
        this.renderClearFiltersButton = this.renderClearFiltersButton.bind(this);
        this.clearFilters = this.clearFilters.bind(this);
        this.state={
            filterResultsInputString: "",
            filtered: false,
            filtered_places: [],
            filterItems: []
        }
    }

    render() {
        if(!this.props.isOpen){
            return null; }
        return (
            <Modal toggle={this.props.hideResultsModal} isOpen={this.props.isOpen} className="">
                <ModalHeader toggle={this.props.hideResultsModal}>Search Results</ModalHeader>
                <ModalBody>
                    <FormGroup>
                        <Label for="filterResults">Filter these results:</Label>
                        <InputGroup>
                            <Input type="text" name="filterResults" id="filterResultsInput" placeholder="Search Within Results (ex. Colorado)"
                                   value={this.state.filterResultsInputString}
                                   onChange={this.setFilterResultsInputString}
                                   onKeyPress={event => {if (event.key === "Enter") {this.filterResults(this.state.filterResultsInputString)}}}/>
                            <InputGroupAddon addonType="append">
                                <ButtonComponent id="filterResultsButton" color="primary" onClick={( ()=> this.filterResults(this.state.filterResultsInputString) )} className="float-right" text='Filter'/>
                            </InputGroupAddon>
                        </InputGroup>
                    </FormGroup>
                    <Container> {this.state.filterItems.map( item => this.createFilterTermsRow(item))}
                        {this.renderClearFiltersButton()}
                    </Container>
                    {this.renderPlaces()}
                </ModalBody> <ModalFooter> </ModalFooter>
            </Modal>
        );}

    createFilterTermsRow(item){
        return(
            <Row key={item}>Filtered result for "{item}"</Row>
        )
    }

    renderClearFiltersButton(){
        if(this.state.filtered){
            return(
                <ButtonComponent id="clearFilters" color="info" onClick={( ()=> this.clearFilters() )} className="float-sm-right" text='Clear Filters'/>
            )
        }
    }

    clearFilters(){
        this.setState({filtered_places:[],filtered:false,filterResultsInputString:"",filterItems:[]})
    }

    renderPlaces(){
        let locationsToDisplay = null;
        if(this.state.filtered){
            locationsToDisplay = this.state.filtered_places;
        } else {
            locationsToDisplay = this.props.places;
        }
        if(this.props.places.length > 0){
            return(
                <Table striped>
                    <thead>
                        <tr>
                            <th>Name</th>
                            <th>Address</th>
                            <th></th>
                        </tr>
                    </thead>
                    <tbody>
                        {locationsToDisplay.map((location, index) => (this.createRow(location, index)))}
                    </tbody>
                </Table>
            );
        } else {
            return "No locations found";
        }
    }

    createRow(location, index){
        let mapClickInfo = {
            "latlng": {
                "lat" : location.latitude,
                "lng" : location.longitude
            },
            "name": location.name,
            "type": location.type
        };
        return(
            <tr key={index}>
                <td>{location.name}</td>
                <td>{location.address}</td>
                <td><Button id='addLocationButton' className="float-right" color="primary"
                            onClick={(() => this.props.addLocation(mapClickInfo))}>Add</Button></td>
            </tr>
        );
    }

    setFilterResultsInputString(onChangeEvent){
        const inputText = onChangeEvent.target.value;
        this.setState({filterResultsInputString: inputText});
    }

    filterResults(inputString){
        let filtered_places_local = [];
        let locationsToFilter = null;
        if(this.state.filtered){
            locationsToFilter = this.state.filtered_places;
        } else {
            locationsToFilter = this.props.places;
        }
        this.setState({filtered: true});

        Object.values(locationsToFilter).forEach( place => {
            Object.values(place).forEach( detail => {
                if(detail.includes(inputString)){
                    filtered_places_local.push(place)
                }
            })
        });
        let newFilterItems = this.state.filterItems
        newFilterItems.push(inputString);
        this.setState({filtered_places:filtered_places_local,filterResultsInputString:"",filterItems:newFilterItems});
    }
}
