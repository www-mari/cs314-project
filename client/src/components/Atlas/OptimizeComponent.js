import React, { Component } from 'react';
import {InputGroup} from 'reactstrap';
import 'leaflet/dist/leaflet.css';
import ButtonComponent from './ButtonComponent';
import {executeOptimizeTrip} from "../../utils/restfulAPI";

export default class OptimizeComponent extends Component {

    constructor(props) {
        super(props);
        this.renderOptimizeButton = this.renderOptimizeButton.bind(this);
        this.addLocation = this.props.addLocation.bind(this);
        this.clearLocationTable = this.props.clearLocationTable.bind(this);
        // this.removeRowandClear = this.props.removeRowandClear.bind(this);

        this.state={
            locations: this.props.locations,
            showMessage: this.props.showMessage,
            optimized_places: [],
        }
    }

    render() {
        return (
            <InputGroup className="mt-2">
                {this.renderOptimizeButton(this.optimizeTrip)}
            </InputGroup>
        );
    }

    renderOptimizeButton(clickToOptimizeTrip){
        return(
                <ButtonComponent
                    id='optimizeButton'
                    color="info"
                    onClick={(() => clickToOptimizeTrip.apply(this) )}
                    text="Optimize"
                />
        );
    }

    async optimizeTrip(){   
        let response = await executeOptimizeTrip(this.state.locations, this.props.serverSettings);     
        this.processTourResponse(response);
    }

    processTourResponse(response){
        this.clearLocationTable();
        //addLocations in reverse order from response.places
        for(let i = 0; i < response.places.length; i++){
            this.addLocation({latlng:{lat:response.places[i].latitude,lng:response.places[i].longitude}})

        }

    }
}
