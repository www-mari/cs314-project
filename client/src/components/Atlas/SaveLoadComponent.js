import React, { Component } from 'react';
import {Button, Input, InputGroup, ButtonGroup} from 'reactstrap';
import 'leaflet/dist/leaflet.css';
import ButtonComponent from "./ButtonComponent";
import SaveModal from './SaveModal';
import LoadModal from './LoadModal';

export default class SaveLoadComponent extends Component {

    constructor(props) {
        super(props);
        this.renderSaveLoadButton = this.renderSaveLoadButton.bind(this);
        this.hideSaveModal= this.hideSaveModal.bind(this);
        this.hideLoadModal= this.hideLoadModal.bind(this);
        this.saveTrip = this.saveTrip.bind(this);
        this.loadTrip = this.loadTrip.bind(this);
        this.renderSaveModal= this.renderSaveModal.bind(this);
        // this.addLocation = this.props.addLocation.bind(this);
        this.state = {
            SaveModalOpen: false,
            LoadModalOpen: false
        }
    }

    render() {
        return (
            <InputGroup className="mt-2" key='saveLoadButton'>
                {this.renderSaveLoadButton(this.saveTrip,this.loadTrip)}
                {this.renderSaveModal()}
                {this.renderLoadModal()}
            </InputGroup>
        );
    }

    renderSaveLoadButton(clickToSaveFunction,clickToLoadFunction){
        return(
            <ButtonGroup>
                <ButtonComponent id='saveButton' color="info" onClick={(() => clickToSaveFunction.apply(this) )} text="Save"/>
                <ButtonComponent id='loadButton' color="info" onClick={(() => clickToLoadFunction.apply(this) )} text="Load"/>
            </ButtonGroup>
        );
    }


    saveTrip(){
        this.setState({SaveModalOpen: true});
    }

    renderSaveModal(){
        return(
            <SaveModal 
                isOpen={this.state.SaveModalOpen}
                hideSaveModal={this.hideSaveModal} 
                locations={this.props.locations}
            />
            
        );
    }

    hideSaveModal(){
        this.setState({SaveModalOpen: false});
    }

    loadTrip(){
        this.setState({LoadModalOpen: true});
    }

    hideLoadModal(){
        this.setState({LoadModalOpen: false});
    }

    renderLoadModal(){
        return(
            <LoadModal 
                isOpen={this.state.LoadModalOpen}
                hideLoadModal={this.hideLoadModal}
                locations={this.props.locations}
                addLocation={this.props.addLocation}
                clearLocations={this.props.removeRowandClear}
            />
        );
    }



}