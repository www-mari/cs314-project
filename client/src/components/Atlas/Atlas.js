import React, {Component, createRef} from 'react';
import {Col, Container, Row, Button} from 'reactstrap';
import {Map, Marker, Popup, TileLayer, Polyline} from 'react-leaflet';
import {getLocation, buildAddress, editAddress, addNotesToLocation} from '../../utils/geocoding';
import Control from 'react-leaflet-control';
import icon from 'leaflet/dist/images/marker-icon.png';
import iconShadow from 'leaflet/dist/images/marker-shadow.png';
import 'leaflet/dist/leaflet.css';
import {isJsonResponseValid, executeFindPlaces, executeFindDistances} from "../../utils/restfulAPI";
import * as findSchema from "../../../schemas/FindResponse.json";
import * as distancesSchema from "../../../schemas/DistancesResponse.json";
import DisplayResultsModal from './DisplayResultsModal';
import SearchComponent from './SearchComponent';
import TableComponent from './TableComponent';
import SaveLoadComponent from './SaveLoadComponent';
const MAP_BOUNDS = [[-90, -180], [90, 180]];
const MAP_CENTER_DEFAULT = L.latLng(40.5734, -105.0865);
const MARKER_ICON = L.icon({iconUrl: icon, shadowUrl: iconShadow, iconAnchor: [12, 40]});
const MAP_LAYER_ATTRIBUTION = "&copy; <a href=&quot;http://osm.org/copyright&quot;>OpenStreetMap</a> contributors";
const MAP_LAYER_URL = "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png";
const MAP_MIN_ZOOM = 1;
const MAP_MAX_ZOOM = 19;

export default class Atlas extends Component {
    constructor(props) {
        super(props);
        this.mapRef = createRef();
        this.constructorBinders();
        this.constructorState();
        this.requestUserLocation();
    }

    constructorBinders(){
        this.addLocation = this.addLocation.bind(this);
        this.handleMapClick = this.handleMapClick.bind(this);
        this.requestUserLocation = this.requestUserLocation.bind(this);
        this.handleGeolocation = this.handleGeolocation.bind(this);
        this.getMarker = this.getMarker.bind(this);
        this.setMarkerPositionDetails = this.setMarkerPositionDetails.bind(this);        
        this.findPlaces = this.findPlaces.bind(this);    
        this.renderResultsModal = this.renderResultsModal.bind(this);
        this.renderSearchComponent = this.renderSearchComponent.bind(this);
        this.handleRemoveRowAndClear = this.handleRemoveRowAndClear.bind(this);
        this.renderPolyline = this.renderPolyline.bind(this);
    }

    constructorState(){
        this.state = {
            mapCenter : MAP_CENTER_DEFAULT,
            markerPosition: null,
            markerPositionDetails: null,
            locations: [], 
            coordinates: {
                inputText: "",
                latLong: null,
                valid: true,
            },
            foundLocations: {
                foundPlaces: null,
                numberFound: null,
                foundSearchInputString: ""
            },
            distances: [],
            showResultsModal: false,
            findLimit: 10
        };
    }

    render() {
        return (
            <div>
                <Container>
                    {this.renderAboutButton()}
                    <Row>
                        <Col sm={12} md={{size: 10, offset: 1}}>
                            {this.renderSearchComponent()}
                            {this.renderLeafletMap()}
                            {this.renderSaveLoadComponent()}
                            {this.renderTableComponent()}                   
                            {this.renderResultsModal()}
                        </Col>
                    </Row>                   
                </Container>
            </div>
        );
    }

    requestUserLocation() {
        if (navigator.geolocation) {
            navigator.geolocation.getCurrentPosition(this.handleGeolocation, console.log("Error retrieving the user's position."));
        }
    }

    handleGeolocation(position) {
        const latlng = {lat: position.coords.latitude, lng: position.coords.longitude};
        this.setState({mapCenter: latlng, markerPosition: latlng});
        this.addLocation({"latlng":latlng}).then();
    }

    renderSaveLoadComponent() {
        return(
            <SaveLoadComponent
                locations={this.state.locations}
                addLocation={this.addLocation}
                removeRowandClear={this.handleRemoveRowAndClear}/>
        )
    }

    renderSearchComponent() {
        return(
            <SearchComponent 
                setCoordinates={(latlng, coordinates) => this.setState({mapCenter: latlng, markerPosition: latlng, coordinates: coordinates})} 
                addLocation={this.addLocation} 
                findPlaces={this.findPlaces} 
                handleSetLimit={(limit) => this.setState({findLimit: limit})}/>
        );
    }

    renderTableComponent() {
        return(
            <TableComponent 
                removeRowandClear={this.handleRemoveRowAndClear}
                markerPositionAndDetails={(markerPos, details) => this.setState({mapCenter: details, markerPosition: markerPos, markerPositionDetails: details})}
                buildAddress={buildAddress}
                distances={this.state.distances}
                locations={this.state.locations}
                showMessage={this.state.showMessage}
                addLocation={this.addLocation}
                serverSettings={this.props.serverSettings}
            />
        );
    }

    handleRemoveRowAndClear(newLocations){
        this.setState({locations: newLocations})
        this.findDistances();
    }

    async findPlaces(findString){
        let response = await executeFindPlaces(findString, this.state.findLimit, this.props.serverSettings);
        if(isJsonResponseValid(response, findSchema)) {
            let locations = this.state.foundLocations;
            locations.foundPlaces = response.places;
            this.setState({foundLocations: locations, showResultsModal: true});
        } else {
            console.log("Atlas.js, processResponse FAILED");
        }
    }

    async findDistances(){
        let response = await executeFindDistances(this.state.locations, this.props.serverSettings);
        if(isJsonResponseValid(response, distancesSchema)) {
            this.setState({distances : response.distances});
        } else {
            console.log("Atlas.js, processDistanceResponse FAILED");
        }
    }

    renderResultsModal(){
        return(
            <DisplayResultsModal 
                isOpen={this.state.showResultsModal}
                hideResultsModal={() => this.setState({showResultsModal: false})}
                addLocation={this.handleMapClick}
                buildAddress={buildAddress}
                getLocation={getLocation}
                places={this.state.foundLocations.foundPlaces}
                foundSearchInputString={this.state.foundLocations.foundSearchInputString}/>
        );
    }

    renderAboutButton(){
        return (
            <Row>
                <Col sm={12} md={{size: 10, offset: 1 }}>
                    <Button id= "toggleAbout" color="primary" block onClick={(() => this.props.toggleAbout())} className="mb-2">
                        View More About Our Team
                    </Button>
                </Col>
            </Row>
        );
    }

    renderFindMeButton(){
        return(
            <Button style={{ background: 'white' }} onClick={this.requestUserLocation} className="mt-2">
                Find Me ðŸ“Œ
            </Button>
        );
    }

    renderPolyline(locations) {
        if(locations.length > 1){
            let positions = [];
            for(let i = 0; i < locations.length; i++){
                positions.push([locations[i].lat, locations[i].lng]);
            }
            positions.push([locations[0].lat, locations[0].lng]);
            return (
                <Polyline color="blue" positions={positions} />
            );
        }
    }

    renderLeafletMap() {
        return (
            <Map
                ref = {this.mapRef}
                className={'mapStyle'}
                boxZoom={false}
                useFlyTo={true}
                zoom={this.getZoom()}
                minZoom={MAP_MIN_ZOOM}
                maxZoom={MAP_MAX_ZOOM}
                maxBounds={MAP_BOUNDS}
                center= {this.state.mapCenter}
                onClick={this.handleMapClick}>
                {this.renderPolyline(this.state.locations)}
                <TileLayer url={MAP_LAYER_URL} attribution={MAP_LAYER_ATTRIBUTION}/>
                    {this.getMarker()}
                <Control position="bottomright">
                    {this.renderFindMeButton()}
                </Control>
            </Map>
        );
    }

    getZoom(){
        if(this.mapRef && this.mapRef.current){
            return this.mapRef.current.leafletElement.getZoom();
        }
        return 15;    
    }

    handleMapClick(mapClickInfo){
        let newMarkerPosition = mapClickInfo.latlng;
        this.addLocation(mapClickInfo);
        this.setState({markerPosition: newMarkerPosition});
    }

    getMarker() {
        if (this.state.markerPosition) {
            let latlngText = "Loading...";
            if(this.state.markerPositionDetails) {
                latlngText = this.state.markerPositionDetails.address;
            } else {
                this.setMarkerPositionDetails(this.state.markerPosition);
            }
            return (
                <Marker ref={(ref) => this.showMarkerPopup(ref)} position={this.state.markerPosition} icon={MARKER_ICON}>
                    <Popup offset={[0, -18]} className="font-weight-bold">
                        {latlngText}
                    </Popup>
                </Marker>
            );
        }
    }

    showMarkerPopup(ref) {
        if (ref) {
            ref.leafletElement.openPopup();
        }
    }

    async setMarkerPositionDetails(latLng) {
        await getLocation(latLng).then(
            (result) => {
                this.setState({markerPositionDetails: result});
            }
        );
    }

    async addLocation(mapClickInfo){
        let newLocations = this.state.locations;
        let location = await getLocation(mapClickInfo.latlng);
        location.name = mapClickInfo.name;
        location.type = mapClickInfo.type;
        if(mapClickInfo.address){
            editAddress(location, mapClickInfo.address)
        }
        if(mapClickInfo.notes){
            addNotesToLocation(location, mapClickInfo.notes)
        }
        newLocations.push(location);
        this.setState({
            locations: newLocations, 
            markerPositionDetails: location, 
            showResultsModal: false, 
            mapCenter: mapClickInfo.latlng
        });

        this.findDistances();
    }
}