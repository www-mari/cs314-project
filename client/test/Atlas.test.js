import './jestConfig/enzyme.config.js';
import {shallow} from 'enzyme';

import React from 'react';
import {create} from "react-test-renderer";
import {Marker} from 'react-leaflet';
import Atlas from '../src/components/Atlas/Atlas';



describe('Atlas', () => {
    const createSnackBar = jest.fn();
    let atlasWrapper;

    beforeEach(() => {
        atlasWrapper = shallow(<Atlas createSnackBar={createSnackBar}/>);
    });

    it('initializes as expected', () => {
        const actualMarkerPosition = atlasWrapper.state().markerPosition;
        const expectedMarkerPosition = null;

        expect(actualMarkerPosition).toEqual(expectedMarkerPosition);
    });

    it('renders a marker on click', () => {
        const actualMarkerPosition = atlasWrapper.state().markerPosition;
        const expectedMarkerPosition = null;

        expect(actualMarkerPosition).toEqual(expectedMarkerPosition);

        const clickPosition = {lat: 0, lng: 0};
        simulateOnClickEvent(atlasWrapper, {latlng: clickPosition});

        expect(atlasWrapper.state().markerPosition).toEqual(clickPosition);
        expect(atlasWrapper.find(Marker).length).toEqual(1);
    });
   
    test("Matches the snapshot", () => {
        const button = create(<Button />);
        expect(button.toJSON()).toMatchSnapshot();
    });

    it('findLimit is default 10', () =>{
        expect(atlasWrapper.state().findLimit).toEqual(10);
    });

    it('requestUserLocation works', () =>{
        atlasWrapper.instance().requestUserLocation(); 
    });

    it('handleGeolocation works', () =>{
        let position= {"coords":[{"latitude":"12"},{"longitude":"12"}] };
        atlasWrapper.instance().handleGeolocation(position); 
    });

    it('handleRemoveRowAndClear works', () =>{
        let NewLocations= ["hi"];
        atlasWrapper.instance().handleRemoveRowAndClear(NewLocations); 
    });

    it('toggleAbout button works', () => {
        let toggleAbout= jest.fn(); 
        atlasWrapper = shallow(<Atlas createSnackBar={createSnackBar} toggleAbout={toggleAbout}/>);

        atlasWrapper.find('#toggleAbout').simulate('click');
    });

    it('renderPolyline works', () =>{
        let mockLocations=[{"lat": 10, "lng": 10}, {"lat": 5, "lng": 5}, {"lat": 15, "lng": 15}];
        atlasWrapper.instance().renderPolyline(mockLocations); 
    });

    function Button(props) {
        return <button>Nothing to do for now</button>;
      }

    function simulateOnClickEvent(wrapper, event) {
        wrapper.find('Map').at(0).simulate('click', event);
        wrapper.update();
    }
});