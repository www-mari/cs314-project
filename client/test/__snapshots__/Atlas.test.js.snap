// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atlas Matches the snapshot 1`] = `
<button>
  Nothing to do for now
</button>
`;
